### PR TITLE
[dag] Plug in the Dag

### LINKED ISSUE #203: [DAG] Feed the DAG on launch
### Story
Our DAG is now ready for production ingestion of certificates. See
https://github.com/MystenLabs/narwhal/tree/main/dag

We want to include entrypoints to consensus in :
https://github.com/MystenLabs/narwhal/blob/main/consensus/src/dag.rs

And run it when `consensus == false`.

And exercise / test:
- ingesting certificates,
- removing them,
- graph traversals

- #203

### PR DESCRIPTION
**Pull request by huitseeker**
This closes #203

### CODE REVIEW

**Review by akichidis — COMMENTED** [review]
A quick pass while waiting for the final version - this looks awesome!

A couple of things:
* I see the `contains_live` method - what about having a method to detect whether a certificate is a weak reference? When we perform causal completion we probably need to have a way to tell whether we have processed already a certificate based on the `DAG` (not the store since we remove certificates from there) [see](https://github.com/MystenLabs/narwhal/blob/b7d2dfdd6394afa31cb910be0ff53efe7c6a5578/primary/src/synchronizer.rs#L141)
* I think it would be useful to have something like a `notify_read` method (like Store has) to get notified once a certificate has been inserted and became available in the DAG. That will help us know that a certificate has reached the DAG during a `read_causal` request with a missing certificate (when we trigger a manual causal completion).

File: consensus/src/dag.rs [hunk 1]
```diff
@@ -1,28 +1,352 @@
...
+/// Represents the errors that can be encountered in this concrete, [`VerifyingKey`],
+/// [`Certificate`] and [`Round`]-aware variant of the Dag.
+#[derive(Debug, Error)]
+pub enum ValidatorDagError<PublicKey: VerifyingKey> {
+    #[error("No remaining certificates for this authority: {0}")]
+    OutOfCertificates(PublicKey),
+    #[error("No known certificates for this authority: {0} at round {1}")]
+    NoCertificateForCoordinates(PublicKey, Round),
+
+    // The generic Dag structure
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply consensus/src/dag.rs:59]: No, this is meant to outline the abstraction boundary between the Dag as we call it here (a Narwhal Dag of certificates that's specific to how things work w.r.t. those) and the Dag as we call it in the dag crate (a graph-theoretic data structure that knows nothing about Certificates).

[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/src/dag.rs:59]: Ah didn't get that, thanks for explaining! I think adding a bit more info could help here

**Review by akichidis — APPROVED** [review]
Excellent work 💯  don't have much to comment. Please see my comment about some extra testing on negative scenarios.

Also let me know your thoughts regarding my earlier comment:

> A couple of things:
> * I see the `contains_live` method - what about having a method to detect whether a certificate is a weak reference? When we perform causal completion we probably need to have a way to tell whether we have processed already a certificate based on the DAG (not the store since we remove certificates from there) [see](https://github.com/MystenLabs/narwhal/blob/b7d2dfdd6394afa31cb910be0ff53efe7c6a5578/primary/src/synchronizer.rs#L141) .
> * I think it would be useful to have something like a `notify_read` method (like Store has) to get notified once a certificate has been inserted and became available in the DAG. That will help us know that a certificate has reached the DAG during a read_causal request with a missing certificate (when we trigger a manual causal completion).

File: consensus/src/dag.rs [hunk 2]
```diff
@@ -1,28 +1,352 @@
...
+        certificate: Certificate<PublicKey>,
+    ) -> Result<(), ValidatorDagError<PublicKey>> {
+        let digest = certificate.digest();
+        let round = certificate.round();
+        let origin = certificate.origin();
+
+        // This fails if the validation of the certificate is incomplete
+        self.dag.try_insert(certificate)?;
+        // TODO: atomicity with the previous
+        self.vertices.insert((origin, round), digest);
```
[the remark this answers is not in the document]
Comment (huitseeker) [reply consensus/src/dag.rs]: Since we extract `(origin, round)` from the certificate itself, and its hash calculation covers over those elements, short of finding a collision on blake2, we will insert the exact same record. I don't see grounds for rejection.
On  the Dag, the re-insertion will be (precisely) a no-op (it won't revive a compressed vertex, for instance).

**Comment by huitseeker** [comment]
> I see the contains_live method - what about having a method to detect whether a certificate is a weak reference? When we perform causal completion we probably need to have a way to tell whether we have processed already a certificate based on the DAG (not the store since we remove certificates from there) [see](https://github.com/MystenLabs/narwhal/blob/b7d2dfdd6394afa31cb910be0ff53efe7c6a5578/primary/src/synchronizer.rs#L141) .

The semantics of whether what you can get out of the store is a weak reference is different than "is the node still in the Dag": a node is still live in the Dag if its reference count is non-zero (i.e. any weak reference is upgradeable). This is what `contains_live` checks. `contains` would tell you if the `certificate digest` matches a past or present Dag vertex, i.e. if the `Certificate` was inserted in the Dag at some point.

I think that the store must remain the source of truth for the availability of a certificate. @akichidis Please open a bug report for deletions from the `certificate_store` you think are illegitimate, we should definitely make sure we check them.


> I think it would be useful to have something like a notify_read method (like Store has) to get notified once a certificate has been inserted and became available in the DAG. That will help us know that a certificate has reached the DAG during a read_causal request with a missing certificate (when we trigger a manual causal completion).

Yup. I opened #233 to track.
